File: src/Homework1/Homework1.Tests/ParserTests.cs
using Xunit;

namespace Homework1.Tests
{
    public class ParserTests
    {
        [Theory]
        [InlineData(new string[] {"2", "+", "3"}, 2, "+", 3, 0)]
        [InlineData(new string[] {"2", "*", "3"}, 2, "*", 3, 0)]
        [InlineData(new string[] {"2", "-", "1"}, 2, "-", 1, 0)]
        [InlineData(new string[] {"2", "/", "2"}, 2, "/", 2, 0)]
        public void TryParseArguments_ValidArguments_ReturnAsExcepted(
            string[] args, int exceptedVal1,
            string exceptedOperation, 
            int exceptedVal2, 
            int excepted)
        {
            var actual = Parser.TryParseArguments(args, out var val1, out var operation, out var val2);
            Assert.Equal(excepted, actual);
            Assert.Equal(exceptedVal1, val1);
            Assert.Equal(exceptedOperation, operation);
            Assert.Equal(exceptedVal2, val2);
        }
        
        [Theory]
        [InlineData(new string[] {".1", "+", ".3"}, 2, "+", 3, 1)]
        public void TryParseArguments_InvalidValues_ReturnAsExcepted(
            string[] args, int exceptedVal1,
            string exceptedOperation, 
            int exceptedVal2, 
            int excepted)
        {
            var actual = Parser.TryParseArguments(args, out var val1, out var operation, out var val2);
            Assert.Equal(excepted, actual);
        }
        
        [Theory]
        [InlineData(new string[] {"2", ".", "3"}, 2, ".", 3, 2)]
        public void TryParseArguments_InvalidOperation_ReturnAsExcepted(
            string[] args, int exceptedVal1,
            string exceptedOperation, 
            int exceptedVal2, 
            int excepted)
        {
            var actual = Parser.TryParseArguments(args, out var val1, out var operation, out var val2);
            Assert.Equal(excepted, actual);
            Assert.Equal(exceptedVal1, val1);
            Assert.Equal(exceptedOperation, operation);
            Assert.Equal(exceptedVal2, val2);
        }
    }
}

File: tests/HW5.Tests/ParserTests.fs
module HW5.Tests.Parser

open Xunit
open HW5.Parser
open HW5.Result


[<Theory>]
[<InlineData("1", "+", "3", 3)>]
let ``checkArgsCount valid`` arg1 arg2 arg3 expected =
    let actual = checkArgsCount [|arg1; arg2; arg3|]
    Assert.Equal(Success expected, actual)
    
[<Theory>]
[<InlineData("1", "+")>]
let ``checkArgsCount invalid with 2 args count`` arg1 arg2=
    let actual = checkArgsCount [|arg1; arg2|]
    Assert.Equal(Fail "Wrong args count", actual)
    
[<Theory>]
[<InlineData("1", 1)>]
[<InlineData("3", 3)>]
[<InlineData("4", 4)>]
let ``parseValue valid`` arg excepted=
    let actual = parseValue arg
    Assert.Equal(Success excepted, actual)
    
    
[<Theory>]
[<InlineData(".")>]
[<InlineData("*")>]
[<InlineData("/")>]
let ``parseValue invalid`` arg=
    let actual = parseValue arg
    Assert.Equal(Fail "Invalid value", actual)
    
[<Theory>]
[<InlineData("+")>]
[<InlineData("-")>]
[<InlineData("*")>]
[<InlineData("/")>]
let ``parseOperand valid`` arg=
    let actual = parseOperand arg
    Assert.Equal(Success arg, actual)
    
[<Theory>]
[<InlineData(".")>]
let ``parseOperand invalid`` arg=
    let actual = parseOperand arg
    Assert.Equal(Fail "Invalid operand", actual)
    